tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/tracee/UJsonTraceReader.scala

```diff
@@ -0,0 +1,94 @@
+package at.forsyte.apalache.tla.tracee
+
+import at.forsyte.apalache.infra.passes.options.SourceOption
+import at.forsyte.apalache.infra.passes.options.SourceOption.FileSource
+import at.forsyte.apalache.io.json.impl.{UJsonRep, UJsonScalaFactory, UJsonToTlaViaBuilder}
+import at.forsyte.apalache.infra.passes.options.SourceOption._
+import at.forsyte.apalache.io.json.JsonDeserializationError
+import at.forsyte.apalache.io.lir.TypeTagReader
+import at.forsyte.apalache.tla.imp.src.SourceStore
+import at.forsyte.apalache.tla.lir.oper.{TlaBoolOper, TlaOper}
+import at.forsyte.apalache.tla.lir.{NameEx, OperEx, TlaEx, TlaOperDecl}
+
+import scala.annotation.unused
+import scala.util.{Failure, Success, Try}
+
+/**
+ * Reads an ITF-format or a TlaJSON-format trace into a TLAIR trace, with an intermediate UJson representation.
+ *
+ * @author
+ *   Jure Kukovec
+ */
+class UJsonTraceReader(sourceStoreOpt: Option[SourceStore], tagReader: TypeTagReader) extends TraceReader[UJsonRep] {
+  private val builder = new UJsonToTlaViaBuilder(sourceStoreOpt)(tagReader)
+  override def read(source: SourceOption): UJsonRep = {
+    val readable: ujson.Readable = source match {
+      case FileSource(f, Format.Json) => f
+      case _                          => throw new IllegalArgumentException("File provided with --trace is not a JSON.")
+    }
+
+    // Rethrow as JsonDeserializationError if unable to read
+    Try(ujson.read(readable)) match {
+      case Success(ujsonVal) => UJsonRep(ujsonVal)
+      case Failure(exception) =>
+        throw new JsonDeserializationError("Unable to read --trace as JSON.", exception)
+    }
+  }
+
+  // TODO
+  private def convertITF(@unused json: UJsonRep): Trace = IndexedSeq.empty
+
+  private def kvFromAsgn(ex: TlaEx): (String, TlaEx) = ex match {
```

Comment:
I didn't know what his mean until I looked at error, so I think the existing name won't read very well in its use site:

```suggestion
  private def kvFromAssignment(ex: TlaEx): (String, TlaEx) = ex match {
```